File: examples/agentic-mentorloop-ts/src/tools/ragTips.ts
import * as fs from 'fs';
import * as path from 'path';
import { TipsRAG } from '../rag/tips';

interface Tip {
  tip_id: string;
  situation: string;
  text: string;
}

interface TipsResult {
  type: 'tips';
  hits: Array<{
    id: string;
    score: number;
    text: string;
    tags: string[];
  }>;
}

interface ErrorResult {
  type: 'error';
  message: string;
  hint: string;
}

export class MentorTips {
  private dataPath: string;
  private rag: TipsRAG;

  constructor(dataPath: string = path.join(__dirname, '../../../data')) {
    this.dataPath = dataPath;
    this.rag = new TipsRAG();
  }

  retrieve(query: string): TipsResult | ErrorResult {
    try {
      const tips = this.loadTips();
      const documents = tips.map(tip => ({
        id: tip.tip_id,
        text: `${tip.situation} ${tip.text}`,
        tags: tip.situation.split(/[,_\s]+/).filter(t => t.length > 0)
      }));
      
      const results = this.rag.findSimilar(query, documents, 2);
      
      if (results.length === 0) {
        return {
          type: 'tips',
          hits: []
        };
      }
      
      // Map results to structured format
      const hits = results.map(r => {
        const tip = tips.find(t => t.tip_id === r.id)!;
        return {
          id: r.id,
          score: parseFloat(r.score.toFixed(2)),
          text: tip.text,
          tags: r.tags || []
        };
      });
      
      return {
        type: 'tips',
        hits
      };
    } catch (error) {
      return {
        type: 'error',
        message: error instanceof Error ? error.message : 'Unknown error occurred',
        hint: 'Check that tips.json exists and is properly formatted'
      };
    }
  }

  private loadTips(): Tip[] {
    const filePath = path.join(this.dataPath, 'tips.json');
    const content = fs.readFileSync(filePath, 'utf-8');
    return JSON.parse(content);
  }
}

File: examples/agentic-mentorloop-py/agentic_mentorloop/tools/rag_tips.py
import json
import os
import re
from typing import List, Dict, Union
from ..rag.tips import TipsRAG


class MentorTips:
    def __init__(self, data_path: str = None):
        if data_path is None:
            data_path = os.path.join(os.path.dirname(__file__), '../../../data')
        self.data_path = data_path
        self.rag = TipsRAG()
    
    def retrieve(self, query: str) -> Dict:
        try:
            tips = self._load_tips()
            documents = [
                {
                    'id': tip['tip_id'],
                    'text': f"{tip['situation']} {tip['text']}",
                    'tags': re.split(r'[,_\s]+', tip['situation'])
                }
                for tip in tips
            ]
            
            results = self.rag.find_similar(query, documents, 2)
            
            if not results:
                return {
                    'type': 'tips',
                    'hits': []
                }
            
            # Map results to structured format
            hits = []
            for r in results:
                tip = next(t for t in tips if t['tip_id'] == r['id'])
                hits.append({
                    'id': r['id'],
                    'score': round(r['score'], 2),
                    'text': tip['text'],
                    'tags': [tag for tag in r.get('tags', []) if tag]
                })
            
            return {
                'type': 'tips',
                'hits': hits
            }
        except Exception as e:
            return {
                'type': 'error',
                'message': str(e),
                'hint': 'Check that tips.json exists and is properly formatted'
            }
    
    def _load_tips(self) -> List[Dict]:
        file_path = os.path.join(self.data_path, 'tips.json')
        with open(file_path, 'r') as f:
            return json.load(f)